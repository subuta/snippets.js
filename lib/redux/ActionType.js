import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'
import pluralize from 'pluralize'

const toActionType = (prefix, type, suffix = '') => {
  if (!type) return _.toUpper(_.snakeCase(prefix))
  return _.toUpper(_.snakeCase(`${prefix}_${type}${suffix && '_' + suffix}`))
}

export const ActionTypeDef = function (...args) {
  const ACTION_TYPE = toActionType.apply(this, args)
  return build`
    const ${ACTION_TYPE} = ${s.stringify(ACTION_TYPE)}
  `
}

export default {
  raw: toActionType,
  set: (model) => toActionType('set', pluralize(model)),
  setIds: (model) => toActionType('set', pluralize.singular(model), 'ids'),
  request: (model) => toActionType('request', pluralize(model)),
  failure: (model) => toActionType('request', pluralize(model), 'failure'),
  create: (model) => toActionType('create', pluralize.singular(model)),
  show: (model) => toActionType('show', pluralize.singular(model)),
  all: (model) => toActionType('index', pluralize(model)),
  update: (model) => toActionType('update', pluralize.singular(model)),
  destroy: (model) => toActionType('delete', pluralize.singular(model)),
}
