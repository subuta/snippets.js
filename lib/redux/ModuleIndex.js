import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'
import pluralize from 'pluralize'

const createModule = (Routes) => {
  const keys = _.map(_.keys(Routes), (k) => pluralize.singular(k))

  const imports = s.import([
    ['redux', null, [
      ['combineReducers']
    ]],
    ['react-router-redux', null, [
      ['routerReducer']
    ]],
    ['lodash', '_'],
    ..._.zip(_.map(keys, k => `./${k}`), keys)
  ])

  const reducers = _.zipObject(keys, keys)

  return build`
    ${imports}
    
    const reducers = ${s.raw({routing: 'routerReducer', ...reducers})}
    
    const makeRootReducer = (injectedReducers) => {
      return combineReducers({
        ...reducers,
        ...injectedReducers
      })
    }
    
    // extract entities from reducers.
    export const fetchEntities = (state) => {
      return _.transform(state, (result, s, key) => {
        if (!s.entities) return
        result[key] = _.get(s, 'entities', {})
      }, {})
    }
    
    // borrowed from https://github.com/davezuko/react-redux-starter-kit/blob/master/src/store/reducers.js
    export const injectReducer = (store, key, reducer) => {
      store.injectedReducers[key] = reducer
      store.replaceReducer(makeRootReducer(store.injectedReducers))
    }
    
    ${s.export('makeRootReducer')}
  `
}

export default createModule