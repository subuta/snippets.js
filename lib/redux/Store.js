import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'
import pluralize from 'pluralize'

export default () => {
  const imports = s.import([
    ['redux', null, [
      ['createStore'],
      ['applyMiddleware'],
      ['compose'],
    ]],
    ['redux-thunk', 'thunk'],
    ['src/utils/history', 'history'],
    ['react-router-redux', null, [
      ['routerMiddleware']
    ]],
    ['./modules', 'makeRootReducer'],
  ])

  return build`
    ${imports}
    
    const middlewares = [thunk, routerMiddleware(history)]
    
    const store = createStore(
      makeRootReducer(),
      compose(
        applyMiddleware(...middlewares),
        window.devToolsExtension ? window.devToolsExtension() : f => f
      )
    )
    
    store.injectedReducers = {};
    
    if (module.hot) {
      module.hot.accept('./modules', () => {
        const makeRootReducer = require('./modules').default
        store.replaceReducer(makeRootReducer(store.injectedReducers))
      })
    }
    
    export default store

  `
}