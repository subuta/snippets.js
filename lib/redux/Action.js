import { build, format, snippets as s } from 'bld.js'

import ActionType from './ActionType'

export default () => {
  // const imports = s.import([
  //   ['lodash', '_'],
  //   ['objection', null, [
  //     'Model'
  //   ]],
  //   ['src/utils/ajvValidator', 'ajvValidator']
  // ])

  return build`
    // action
    console.log('action')
  `
}
