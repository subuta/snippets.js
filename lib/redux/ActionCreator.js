import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'
import pluralize from 'pluralize'

import createAction from './Action'
import ActionType from 'lib/redux/ActionType'

const set = ({schema}) => {
  const {tableName} = schema
  // ensure naming convention.
  const model = _.camelCase(pluralize.singular(tableName))

  const ACTION_TYPE = ActionType.set(model)
  const name = _.camelCase(ACTION_TYPE)
  const Action = createAction(ACTION_TYPE, tableName)

  return build`
    const ${name} = (${tableName}) => {
      return ${Action}
    }
  `
}

const setIds = ({schema}) => {
  const {tableName} = schema
  // ensure naming convention.
  const model = _.camelCase(pluralize.singular(tableName))

  const ACTION_TYPE = ActionType.setIds(model)
  const name = _.camelCase(ACTION_TYPE)
  const Action = createAction(ACTION_TYPE, 'ids')

  return build`
    const ${name} = (ids) => {
      return ${Action}
    }
  `
}

const create = ({schema}) => {
  const {tableName} = schema

  // ensure naming convention.
  const model = _.camelCase(pluralize.singular(tableName))

  const ACTION_TYPE = ActionType.create(model)
  const name = _.camelCase(ACTION_TYPE)

  const SET_ACTION_TYPE = ActionType.set(model)
  const setAction = _.camelCase(SET_ACTION_TYPE)

  return build`
    const ${name} = (params) => {
      return (dispatch) => {
        dispatch({type: ${ActionType.request(model)}})
        return api.${model}.create(params).then((data) => {
          /* mat Create data transform [start] */
          /* mat Create data transform [end] */
          dispatch(${setAction}(normalize(data, ${model})))
          return data
        })        
      }
    }
  `
}

const show = ({schema}) => {
  const {tableName} = schema

  // ensure naming convention.
  const model = _.camelCase(pluralize.singular(tableName))

  const ACTION_TYPE = ActionType.request(model)
  const name = _.camelCase(ACTION_TYPE)

  const SET_ACTION_TYPE = ActionType.set(model)
  const setAction = _.camelCase(SET_ACTION_TYPE)

  return build`
    const ${pluralize.singular(name)} = (id) => {
      return (dispatch) => {
        dispatch({type: ${ActionType.request(model)}})
        return api.${model}.show(id).then((data) => {
          /* mat Show data transform [start] */
          /* mat Show data transform [end] */
          dispatch(${setAction}(normalize(data, ${model})))
          return data
        })        
      }
    }
  `
}

const all = ({schema}) => {
  const {tableName} = schema

  // ensure naming convention.
  const model = _.camelCase(pluralize.singular(tableName))

  const ACTION_TYPE = ActionType.request(model)
  const name = _.camelCase(ACTION_TYPE)

  const SET_ACTION_TYPE = ActionType.set(model)
  const setAction = _.camelCase(SET_ACTION_TYPE)

  return build`
    const ${name} = () => {
      return (dispatch) => {
        dispatch({type: ${ActionType.request(model)}})
        return api.${model}.index().then((data) => {
          /* mat Index data transform [start] */
          /* mat Index data transform [end] */
          dispatch(${setAction}(normalize(data, ${model}List)))
          return data
        })        
      }
    }
  `
}

const update = ({schema}) => {
  const {tableName} = schema

  // ensure naming convention.
  const model = _.camelCase(pluralize.singular(tableName))

  const ACTION_TYPE = ActionType.update(model)
  const name = _.camelCase(ACTION_TYPE)

  const SET_ACTION_TYPE = ActionType.set(model)
  const setAction = _.camelCase(SET_ACTION_TYPE)

  return build`
    const ${pluralize.singular(name)} = (id, params) => {
      return (dispatch) => {
        dispatch({type: ${ActionType.request(model)}})
        return api.${model}.update(id, params).then((data) => {
          /* mat Update data transform [start] */
          /* mat Update data transform [end] */
          dispatch(${setAction}(normalize(data, ${model})))
          return data
        })        
      }
    }
  `
}

const destroy = ({schema}) => {
  const {tableName} = schema

  // ensure naming convention.
  const model = _.camelCase(pluralize.singular(tableName))

  const ACTION_TYPE = ActionType.destroy(model)
  const name = _.camelCase(ACTION_TYPE)

  const SET_ACTION_TYPE = ActionType.set(model)
  const setAction = _.camelCase(SET_ACTION_TYPE)

  const SET_IDS_ACTION_TYPE = ActionType.setIds(model)
  const setIdsAction = _.camelCase(SET_IDS_ACTION_TYPE)

  return build`
    const ${pluralize.singular(name)} = (id) => {
      return (dispatch, getState) => {
        dispatch({type: ${ActionType.request(model)}})
        return api.${model}.destroy(id).then(() => {
          const state = getState()
          const nextIds = _.without(getIds(state), id)
          dispatch(${setIdsAction}(nextIds))
        })        
      }
    }
  `
}

export default {
  set,
  setIds,
  create,
  show,
  all,
  update,
  destroy
}
