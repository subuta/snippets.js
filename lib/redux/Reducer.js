import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'

import ActionType from './ActionType'
import pluralize from 'pluralize'
import parseRelations from 'lib/_utils/parseRelations'

export const entities = (Model) => {
  const {schema} = Model
  const {tableName} = schema
  // ensure naming convention
  const name = _.camelCase('entities')
  const model = _.camelCase(pluralize.singular(tableName))

  return build`
    const ${name} = (state = {}, action) => {
      if (_.get(action, ['payload', 'entities', '${model}'])) {
        return {...state, ...action.payload.entities.${model}}
      }
      return state
    }
  `
}

export const ids = ({schema}) => {
  const {tableName} = schema
  // ensure naming convention
  const name = _.camelCase('ids')
  const model = _.camelCase(tableName)

  return build`
    const ${name} = (state = [], action) => {
      if (action.type === ${ActionType.set(model)}) {
        if (_.isArray(action.payload.result)) {
          return _.uniq([...state, ...action.payload.result])
        }
        return _.uniq([...state, action.payload.result])
      }
      return state
    }
  `
}

export const isRequestProgress = ({schema}) => {
  const {tableName} = schema
  // ensure naming convention
  const model = _.camelCase(tableName)

  return build`
    const isRequestProgress = (state = false, action) => {
      if (action.type === ${ActionType.request(model)}) {
        return true
      } else if (action.type === ${ActionType.set(model)} ||
                 action.type === ${ActionType.failure(model)}) {
        return false
      }
      return state
    }
  `
}

export default {
  entities,
  ids,
  isRequestProgress
}
