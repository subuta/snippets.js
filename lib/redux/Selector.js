import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'
import pluralize from 'pluralize'

import Reducer from './Reducer'

const selectorFor = ({ tableName }, name) => {
  // ensure naming convention
  name = _.camelCase(name)

  const model = _.camelCase(pluralize.singular(tableName))
  const get = _.camelCase('get' + _.upperFirst(name))

  return `const ${get} = state.${model}.${name}`
}

const getAll = ({ tableName }) => {
  // ensure naming convention
  tableName = _.camelCase(pluralize.singular(tableName))

  return build`
  const getAll = createSelector(
    getEntities,
    getIds,
    _.identity,
    (entities, ids, state) => ids.map(id => {
      return denormalize(entities[id], '${tableName}', state)
    })
  )
`
}

// Generate selector for reducers based on named exports.
let selectors = _.transform(_.keys(Reducer), (result, nameOfReducer) => {
  result[nameOfReducer] = (model) => selectorFor(model, nameOfReducer)
}, {})

selectors.getAll = getAll

export default selectors