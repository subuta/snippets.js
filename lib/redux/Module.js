import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'
import pluralize from 'pluralize'

import ActionType, { ActionTypeDef } from './ActionType'
import ActionCreator from './ActionCreator'
import Reducer from './Reducer'
import Selector from './Selector'
import getControllerActions from 'lib/_utils/getControllerActions'
import parseRelations from 'lib/_utils/parseRelations'

const createModule = (RouteConfig, ModelConfig) => {
  const {
    schema
  } = ModelConfig
  let {tableName} = schema

  const keys = getControllerActions(RouteConfig)

  const renderAction = (action) => {
    switch (action) {
      case 'index':
        return ActionCreator.all(ModelConfig) + s.EOL
      case 'show':
        return ActionCreator.show(ModelConfig) + s.EOL
      case 'create':
        return ActionCreator.create(ModelConfig) + s.EOL
      case 'update':
        return ActionCreator.update(ModelConfig) + s.EOL
      case 'destroy':
        return ActionCreator.destroy(ModelConfig) + s.EOL
    }
  }

  const actions = _.map(keys, renderAction)

  // ensure naming convention.
  tableName = _.camelCase(pluralize(tableName))
  const model = pluralize.singular(tableName)

  const relations = parseRelations(ModelConfig)

  let importsFromRelations = s.import(_.map(_.values(relations), (r) => {
    // ensure naming convention.
    r = _.camelCase(pluralize.singular(r))
    return [`./${r}`, null, [
      ActionType.set(r)
    ]]
  }))

  const imports = s.import([
    ['lodash', '_'],
    ['redux', null, [
      'combineReducers'
    ]],
    ['reselect', null, [
      'createSelector'
    ]],
    ['normalizr', null, [
      ['normalize']
    ]],
    ['src/views/utils/schema', null, [
      [model],
      [`${model}List`],
      'denormalize'
    ]],
    ['src/views/utils/api', 'api']
  ])

  return build`
    ${imports}
    
    ${importsFromRelations}
    
    // -------------
    // Constants
    // -------------
    export ${ActionTypeDef(ActionType.request(tableName))}
    export ${ActionTypeDef(ActionType.failure(tableName))}
    export ${ActionTypeDef(ActionType.set(tableName))}
    export ${ActionTypeDef(ActionType.setIds(tableName))}
    
    /* mat Custom constants [start] */
    /* mat Custom constants [end] */
    
    // -------------
    // ActionCreators
    // -------------
    export ${ActionCreator.set(ModelConfig)}
    
    export ${ActionCreator.setIds(ModelConfig)}
    
    ${_.map(actions, (action) => `export ${action}`)}
    
    /* mat Custom actionCreators [start] */
    /* mat Custom actionCreators [end] */
    
    // -------------
    // Reducers
    // -------------
    export ${Reducer.entities(ModelConfig)}
    
    export ${Reducer.ids(ModelConfig)}
    
    export ${Reducer.isRequestProgress(ModelConfig)}
    
    let reducers = {
      entities,
      ids,
      isRequestProgress
    }
    
    /* mat Custom reducers [start] */
    /* mat Custom reducers [end] */
    
    export default combineReducers(reducers)
    
    // -------------
    // Selectors
    // -------------
    export ${Selector.entities(ModelConfig)}
    export ${Selector.ids(ModelConfig)}
    export ${Selector.isRequestProgress(ModelConfig)}
    export ${Selector.getAll(ModelConfig)}
    
    /* mat Custom selectors [start] */
    /* mat Custom selectors [end] */
  `
}

export default createModule
