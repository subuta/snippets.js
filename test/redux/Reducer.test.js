import test from 'ava'
import { build, format, snippets as s } from 'bld.js'

import * as Reducer from 'lib/redux/Reducer'

test.beforeEach(async (t) => {

})

test.afterEach((t) => {
})

test('should create redux entities Reducer', async (t) => {
  const code = Reducer.entities({
    tableName: 'books'
  })

  const expected = build`
    const entities = (state = {}, action) => {
      if (action.type === SET_BOOKS) {
        return {...state, ...action.payload.entities.book}
      }
      return state
    }
  `

  t.is(format(code), format(expected))
})

test('should create redux entities Reducer with relations', async (t) => {
  const code = Reducer.entities({
    tableName: 'books',
    relations: {
      users: {},
      shops: {}
    }
  })

  const expected = build`
    const entities = (state = {}, action) => {
      if (
        action.type === SET_BOOKS ||
        action.type === SET_USERS ||
        action.type === SET_SHOPS
      ) {
        return {...state, ...action.payload.entities.book}
      }
      return state
    }
  `

  t.is(format(code), format(expected))
})

test('should create redux ids Reducer', async (t) => {
  const code = Reducer.ids({
    tableName: 'books'
  })

  const expected = build`
    const ids = (state = [], action) => {
      if (action.type === SET_BOOKS) {
        if (_.isArray(action.payload.result)) {
          return _.uniq([...state, ...action.payload.result])
        }
        return _.uniq([...state, action.payload.result])
      }
      return state
    }
  `

  t.is(format(code), format(expected))
})

test('should create redux requestProgress Reducer', async (t) => {
  const code = Reducer.requestProgress({
    tableName: 'books'
  })

  const expected = build`
    const isRequestProgress = (state = false, action) => {
      if (action.type === REQUEST_BOOKS) {
        return true
      } else if (
        action.type === SET_BOOKS ||
        action.type === REQUEST_BOOKS_FAILURE
      ) {
        return false
      }
      return state
    }
  `

  t.is(format(code), format(expected))
})