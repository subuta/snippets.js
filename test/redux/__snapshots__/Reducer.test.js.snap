// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux/Reducer should create redux entities Reducer 1`] = `
"const entities = (state = {}, action) => {
  if (_.get(action, ['payload', 'entities', 'book'])) {
    return {...state, ...action.payload.entities.book}
  }
  return state
}
"
`;

exports[`redux/Reducer should create redux entities Reducer with relations 1`] = `
"const entities = (state = {}, action) => {
  if (_.get(action, ['payload', 'entities', 'book'])) {
    return {...state, ...action.payload.entities.book}
  }
  return state
}
"
`;

exports[`redux/Reducer should create redux ids Reducer 1`] = `
"const ids = (state = [], action) => {
  if (action.type === SET_BOOKS) {
    if (_.isArray(action.payload.result)) {
      return _.compact(_.uniq([...state, ...action.payload.result]))
    }
    return _.compact(_.uniq([...state, action.payload.result]))
  } else if (action.type === SET_BOOK_IDS) {
    return _.compact(_.uniq(action.payload))
  }
  return state
}
"
`;

exports[`redux/Reducer should create redux requestProgress Reducer 1`] = `
"const isRequestProgress = (state = false, action) => {
  if (action.type === REQUEST_BOOKS) {
    return true
  } else if (
    action.type === SET_BOOKS ||
    action.type === REQUEST_BOOKS_FAILURE
  ) {
    return false
  }
  return state
}
"
`;
